notmuch fails to build on master

### Describe the bug
The build is failing on master; it has been reproduced at c9248a6d87ce071fc7538dc5358b60ce86092e49.

An example of the failure is below:
```console
T670-duplicate-mid: Testing duplicate message ids
 PASS   at most 1 thread-id per xapian document
 PASS   search: first indexed subject preserved
 BROKEN First subject preserved in notmuch-show (json)
        --- T670-duplicate-mid.3.expected       2022-01-10 15:49:39.334299463 +0000
        +++ T670-duplicate-mid.3.output 2022-01-10 15:49:39.334299463 +0000
        @@ -2,6 +2,7 @@
             [
                 [
                     {
        +                "crypto": {},
                         "date_relative": "2001-01-05",
                         "excluded": false,
                         "filename": [
        @@ -12,7 +13,7 @@
                         "headers": {
                             "Date": "GENERATED_DATE",
                             "From": "Notmuch Test Suite <test_suite@notmuchmail.org>",
        -                    "Subject": "message 1",
        +                    "Subject": "message 0",
                             "To": "Notmuch Test Suite <test_suite@notmuchmail.org>"
                         },
                         "id": "XXXXX",
 PASS   Search for second subject
 BROKEN Regexp search for second subject
        --- T670-duplicate-mid.5.EXPECTED       2022-01-10 15:49:39.352299871 +0000
        +++ T670-duplicate-mid.5.OUTPUT 2022-01-10 15:49:39.352299871 +0000
        @@ -1,3 +0,0 @@
        -MAIL_DIR/copy0
        -MAIL_DIR/copy1
        -MAIL_DIR/copy2
 PASS   search for body in duplicate file
 PASS   reindex drops terms in duplicate file
 PASS   reindex choses subject from first filename
 PASS   Deleted first duplicate file does not stop notmuch show from working
...

Notmuch test suite complete.
1537/1559 tests passed.
6 broken tests failed as expected.
16 tests failed.
All tests in 2 files skipped.
ERROR: some tests failed; first failed returned error code 1
make: *** [test/Makefile.local:75: test] Error 1
```

There is a new release (`0.34.3`) which also doesn't pass, so I didn't send the PR as it'd fail on CI.

### Steps To Reproduce
Steps to reproduce the behavior:
1. `nix-build -A notmuch`

### Notify maintainers

@flokli
@puckipedia

--- a/pkgs/applications/networking/mailreaders/notmuch/default.nix
+++ b/pkgs/applications/networking/mailreaders/notmuch/default.nix
@@ -18,6 +18,11 @@ stdenv.mkDerivation rec {
     sha256 = "wfLO7kf2iXESItcgWvKj/npKnYwy5OCyStZviN9qR9M=";
   };
 
+  patches = [
+    # https://nmbug.notmuchmail.org/nmweb/show/87o84iza9r.fsf%40starbuck.i-did-not-set--mail-host-address--so-tickle-me
+    ./test-fix-support-for-gpgsm-in-gnupg-2.3.patch
+  ];
+
   nativeBuildInputs = [
     pkg-config
     doxygen                   # (optional) api docs

--- a/pkgs/applications/networking/mailreaders/notmuch/test-fix-support-for-gpgsm-in-gnupg-2.3.patch
+++ b/pkgs/applications/networking/mailreaders/notmuch/test-fix-support-for-gpgsm-in-gnupg-2.3.patch
@@ -1,0 +1,28 @@
+From a642ad542e3d3f34e949c5c66923ca8a6e6cbbd8 Mon Sep 17 00:00:00 2001
+From: Stig Palmquist <stig@stig.io>
+Date: Tue, 11 Jan 2022 13:23:13 +0100
+Subject: [PATCH] test: fix support for gpgsm in gnupg 2.3
+
+gpgsm --list-keys output changed the label for fingerprints from
+"fingerprint: " to "sha[12] fpr: " breaking tests with gnupg 2.3. this
+adds support for both.
+---
+ test/test-lib.sh | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/test/test-lib.sh b/test/test-lib.sh
+index 6bc0b723..3de608f9 100644
+--- a/test/test-lib.sh
++++ b/test/test-lib.sh
+@@ -145,7 +145,7 @@ add_gpgsm_home () {
+     mkdir -p -m 0700 "$GNUPGHOME"
+     gpgsm --batch --no-tty --no-common-certs-import --pinentry-mode=loopback --passphrase-fd 3 \
+ 	  --disable-dirmngr --import  >"$GNUPGHOME"/import.log 2>&1 3<<<'' <$NOTMUCH_SRCDIR/test/smime/0xE0972A47.p12
+-    fpr=$(gpgsm --batch --list-key test_suite@notmuchmail.org | sed -n 's/.*fingerprint: //p')
++    fpr=$(gpgsm --batch --list-key test_suite@notmuchmail.org | sed -n 's/.*\(fingerprint\|sha1 fpr\): //p')
+     echo "$fpr S relax" >> "$GNUPGHOME/trustlist.txt"
+     gpgsm --quiet --batch --no-tty --no-common-certs-import --disable-dirmngr --import < $NOTMUCH_SRCDIR/test/smime/ca.crt
+     echo "4D:E0:FF:63:C0:E9:EC:01:29:11:C8:7A:EE:DA:3A:9A:7F:6E:C1:0D S" >> "$GNUPGHOME/trustlist.txt"
+-- 
+2.34.1
+

